eth_account/hdaccount/deterministic.py

```diff
@@ -0,0 +1,211 @@
+"""
+Heirarchical Deterministic Wallet generator (HDWallet)
+
+Partially implements the BIP-0032, BIP-0043, and BIP-0044 specifications:
+BIP-0032: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+BIP-0043: https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki
+BIP-0044: https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki
+
+Skips serialization and public key derivation as unnecssary for this library's purposes.
+
+Notes
+-----
+
+* Integers are modulo the order of the curve (referred to as n).
+* Addition (+) of two coordinate pair is defined as application of the EC group operation.
+* Concatenation (||) is the operation of appending one byte sequence onto another.
+
+
+Definitions
+-----------
+
+* point(p): returns the coordinate pair resulting from EC point multiplication
+  (repeated application of the EC group operation) of the secp256k1 base point
+  with the integer p.
+* ser_32(i): serialize a 32-bit unsigned integer i as a 4-byte sequence,
+  most significant byte first.
+* ser_256(p): serializes the integer p as a 32-byte sequence, most significant byte first.
+* ser_P(P): serializes the coordinate pair P = (x,y) as a byte sequence using SEC1's compressed
+  form: (0x02 or 0x03) || ser_256(x), where the header byte depends on the parity of the
+  omitted y coordinate.
+* parse_256(p): interprets a 32-byte sequence as a 256-bit number, most significant byte first.
+
+"""
+# Additional notes:
+# - This module currently only implements private parent key => private child key CKD function,
+#   as it is not necessary to the HD key derivation functions used in this library to implement
+#   the other functions yet (as this module is only used for derivation of private keys). That
+#   could change, but wasn't deemed necessary at the time this module was introduced.
+# - Unlike other libraries, this library does not use Bitcoin key serialization, because it is
+#   not intended to be ultimately used for Bitcoin key derivations. This presents a simplified
+#   API, and no expectation is given for `xpub/xpriv` key derivation.
+from typing import (
+    Tuple,
+    Union,
+)
+
+from eth_utils import (
+    to_int,
+)
+
+from ._utils import (
+    SECP256K1_N,
+    ec_point,
+    hmac_sha512,
+)
+
+
+class Node(int):
+    TAG = ""  # No tag
+    OFFSET = 0x0  # No offset
+    """
+    Base node class
+    """
+    def __new__(cls, index):
+        obj = int.__new__(cls, index + cls.OFFSET)
+        obj.index = index
+        return obj
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.index})"
+
+    def __add__(self, other: int):
+        return self.__class__(self.index + other)
+
+    def serialize(self) -> bytes:
+        assert 0 <= self < 2**32
```

Comment:
What do you think about moving this validation to `__new__` so that we can prevent an invalid one of these from even being instantiated?

Also, we generally stay away from `assert thing` in runtime code because of [how easy it is to disable them at runtime](https://stackoverflow.com/questions/1273211/disable-assertions-in-python).  Using `ValueError` is probably acceptable here if you want to stick with standard library, or `eth_utils.ValidationError` is a common convention across our libraries as well.

Reply:
I'll fix this one up, I did `assert` elsewhere just as an extra size check, even though badly-sized input shouldn't be used outside of this library